savaslabs/sumac
# Issue #6238: Duplicate time entries and exclude clients from spell checking

This PR does a few things:

- Should correctly implement the logic to exclude all time entries logged to a Client from spell checking. See #25 and #26 for some history on this, and commit 770f978744 where it's implemented in this PR
- Start refactoring code so that commands don't have to duplicate logic for loading configuration. Much more can be done here but it's a start.
- Add a command for identifying duplicate time entries
- Add a command for removing duplicate time entries (based on output of the above command)
- Add unit tests
- Run code standards checks and unit tests in Travis

---

To test:

- Bring up our Redmine instance locally with the latest DB by running `make install`. (You might want to do `make clean` first)
- Check out this branch locally. Create a config.yml file if you don't have one already. Double-check that the Redmine URL is `local.pm.savaslabs.com` :smile:
- Review the updated example config.yml provided by this PR and add an entry for `dont_spell_check` if you'd like to test that out. `3228941` is the ID of the "Savas Labs" client.

Try running `php sumac.php`, you should see two new commands.

Try `php sumac.php sync:find-duplicates -s`, this takes a while. Select the output and then run `pbpaste | jq` if you'd like to look at it in a pretty format. You can also run this command without `-s` to get more verbose data.

Now run `php sumac.php sync:remove-duplicates $(pbpaste)`. (note you could also do something like `php sumac.php sync:remove-duplicates $(php sumac.php sync:find-duplicates -s)` but I haven't tested that). The most recent time entry should be kept and all older duplicates should be removed.

If you run `php sumac.php sync:find-duplicates` you should get back an empty array.

Now run `php sumac.php sync:entries 20180411:20180511` and/or `php sumac.php sync:entries --update 20180411:20180511` a few times, you should not see any duplicates on for example issue 6106. Tip: Add `10042639` and `16789765` as IDs in the `debug_projects:` section of the `config.yml`

---

Deployment

- Once this is merged, we'd create two new Jenkins jobs that are freestyle and not run on a schedule.
  - find duplicates job, which an admin can run to see if there are any in the production system
  - remove duplicates job, which would run find-duplicates and pipe the output into the remove duplicates command.
